src/geometry/CTiglBSplineAlgorithms.cpp

```diff
@@ -836,6 +836,194 @@ CTiglApproxResult CTiglBSplineAlgorithms::reparametrizeBSplineNiceKnots(Handle(G
+
+    // u needs to be in range of paramsOld
+    if (u < paramsOld[0] || paramsOld[paramsOld.size()-1] < u)
+        throw CTiglError("Argument of reparameterization function out of range in calcReparamfctInv");
+
+    // First: Find interval [p_i,p_{i+1}] to determine boundaries
+    // a_x = lowerBound, b_x = upperBound
+    double a_x = paramsOld[0];
+    double b_x = paramsOld[1];
+    double a_y = paramsNew[0];
+    double b_y = paramsNew[1];
+
+    for (int paramsIdx=1; paramsIdx < paramsOld.size()-1; paramsIdx++) {
+        // Break condition
+        if (a_x <= u && u <= b_x)
+            paramsIdx = paramsOld.size();
+        else {
+            a_x = paramsOld[paramsIdx];
+            b_x = paramsOld[paramsIdx+1];
+            a_y = paramsNew[paramsIdx];
+            b_y = paramsNew[paramsIdx+1];
+        }
+    }
+
+    // Calc slope m
+    double m = (b_y-a_y) / (b_x-a_x);
+    // Calc intercept n
+    double n = a_y - a_x*m;
+
+    // Calc inverse g(u) [linear function value]
+    return m*u + n;
+}
+
+TColStd_Array1OfReal calcNewKnotVectorS(const Handle(Geom_BSplineCurve) curve, std::vector<double> paramsOld, std::vector<double> paramsNew)
+{
+    // This function matches part one of the second step of the reparameterization algorithm in The NURBS Book (2nd edition), p. 251
+
+    double newKnot;
+
+    // Set up knot vector S' consisting of s_i=g(u_i) [inverse reparametrization fct]
+    Standard_Integer nbKnots = curve->NbKnots(); // Size of distinct knots
+    TColStd_Array1OfReal newKnots(1, nbKnots);
+    for (int knotIdx = 1; knotIdx <= nbKnots; knotIdx++) {
+        // Calc s_i=g(u_i) of picewise linear fct
+        // defined as inverse of interpolation (x_i,y_i)=(paramsNew[i], paramsOld[i])
+        newKnot = calcReparamfctInv(paramsOld, paramsNew, curve->Knot(knotIdx));
+        newKnots.SetValue(knotIdx, newKnot);
+    }
+    return newKnots;
+}
+
+TColStd_Array1OfInteger setNewKnotMultsS(const Handle(Geom_BSplineCurve) curve)
+{
+    // This function matches part two of the second step of the reparameterization algorithm in The NURBS Book (2nd edition), p. 251
+
+    // Set up mults for new knot vector S'
+    Standard_Integer mult;
+    Standard_Integer nbKnots = curve->NbKnots(); // Size of distinct knots
+    TColStd_Array1OfInteger knotMults(1, nbKnots);
+    for (int knotIdx = 1; knotIdx <= nbKnots; knotIdx++) {
+        // Account for potentially higher mult at global boundaries (usually p+1)
+        // That is why, not the degree is used as mult (as described in the NURBS book), but the mult of the old knot vector
+        // mult should be equal to p*q (degree of original BSpline [p] times degree of reparam fct [q])
+        // We use q=1 for simplification and for not raising the degree
+        // Refined knot vector was set to have m_i=p (or p+1 at global boundaries), so we use it here
+        mult = curve->Multiplicity(knotIdx);
+        knotMults.SetValue(knotIdx, mult);
+    }
+    return knotMults;
+}
```

Comment:
Can't we just use `Geom_BSplineCurve::Multiplicities` instead of `setNewKnotMultsS`? https://dev.opencascade.org/doc/refman/html/class_geom___b_spline_curve.html#a7ada148c6461e1adbd8c71bcef2a4a43

Reply:
Oh yes, you are completely right! I will delete the `setNewKnotMultsS` and directly use the multiplicities